Fix the data sharding for TP.

wang2yn84 opened the pull request:
# Description
Another attempt after rolled back PR#1533 because TPU hardware topology and dcn/ici concept is different from GPU and it breaks the TPU dataloading.

This PR takes another easier approach that it reshards input data from (B,) to (B, S) instead of attempting to load every data shards correctly at the very beginning. It avoids the complicated logic of manual sharding input data at the cost of 1 additional reshard for every batch of data.

# Tests

Run unit tests locally.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

wang2yn84 commented on MaxText/train.py:
```diff
@@ -851,6 +856,8 @@ def train_loop(config, state=None):
       record_goodput(recorder, config, recorder.record_data_loading_start_time if recorder else None)
       try:
         example_batch = load_next_batch(data_iterator, example_batch, config)
+        # Reshard data from (B,) to (B, S)
```
Updated, thank you!

RissyRan commented on MaxText/configs/base.yml:
```diff
@@ -336,6 +336,7 @@ logical_axis_rules: [
                     ]
 # Axes used for DCN must be earlier in this list than ICI, see (b/339009148) for details
 data_sharding: [['data', 'stage', 'fsdp', 'fsdp_transpose', 'sequence', 'context', 'context_autoregressive', 'tensor', 'tensor_transpose', 'tensor_sequence', 'expert', 'autoregressive']]
+input_data_sharding_logical_axes: ['activation_embed_and_logits_batch', 'activation_norm_length']
```
Trying to understand more how do you select those 2 axes? i.e. why not adding `activation_length`, `activation_batch`?

wang2yn84 replied:
activation_embed_and_logits_batch is super set of activation_length and activation_norm_length is super set of activation_batch. These two are selected to include all the mesh axes that we should shard for input data batch and sequence dimension.

RissyRan reviewed (commented):
Thanks for the PR! I see you mentioned `Run unit tests locally.` could you explain a little bit more? i.e. if you have printed out the data loading is sharded expected on each host?

aireenmei commented:
I suggest @A9isha to take a look, she is working on context parallelism on TPU

wang2yn84 commented on MaxText/tests/integration_tests/train_tests.py:
```diff
@@ -236,14 +236,34 @@ def test_gpu_context_parallelism(self):
         "enable_checkpointing=False",
         "attention=cudnn_flash_te",
-        "ici_fsdp_parallelism=2",
+        "ici_fsdp_parallelism=-1",
```
I'm trying to run this test locally with 8 GPUs and hardcoded FSDP gives me trouble. The correct setting should allow both running locally and on presubmit server.

wang2yn84 commented:
> Thanks for the PR! I see you mentioned `Run unit tests locally.` could you explain a little bit more? i.e. if you have printed out the data loading is sharded expected on each host?

I run both tests and tried training steps locally. Since we are changing the input sharding for train_step, if the data doesn't shard as expected, it will end up erroring out.

aireenmei commented:
To be consistent, should we update other places where config.data_sharding is used? Such as in input_pipeline_interface.py, grpo_input_pipeline.py and tests/*_data_processing_test.py.

wang2yn84 commented:
> To be consistent, should we update other places where config.data_sharding is used? Such as in input_pipeline_interface.py, grpo_input_pipeline.py and tests/*_data_processing_test.py.

Hi Aireen, the current solution doesn't change the data sharding when loading the data. Instead, I'm sharding the data after it's loaded in train.py

> To be consistent, should we update other places where config.data_sharding is used? Such as in input_pipeline_interface.py, grpo_input_pipeline.py and tests/*_data_processing_test.py.

Goot catch. I updated the other files that example_batch are feed into the training function. Updated!

aireenmei commented:
Thanks for the update. I guess my question was that if config.data_sharding is still used for data sharding? If not, should we retire this flag? Looks like the remaining place it's not used for sharding purpose. Or we document the usage due to the name confusion.

wang2yn84 commented:
> Thanks for the update. I guess my question was that if config.data_sharding is still used for data sharding? If not, should we retire this flag? Looks like the remaining place it's not used for sharding purpose. Or we document the usage due to the name confusion.

Hi Aireen, yes are still using config.data_sharding for the initial sharding when we first load the data. Then we reshard the data based on input_data_sharding_logical_axes. So config.data_sharding is still necessary.

aireenmei reviewed (approved):
Thanks for replying to the comments!
